File: cpp/TypeScript/CPP14ParserBase.ts
import {Parser, Lexer, Token, TokenStream, ParserRuleContext} from "antlr4";
import CPP14Parser from './CPP14Parser';
import ParametersAndQualifiersContext from './CPP14Parser';

export default abstract class CPP14ParserBase extends Parser {

    constructor(input: TokenStream) {
        super(input);
    }

    protected IsPureSpecifierAllowed() : boolean {
        try {
            var x = this._ctx; // memberDeclarator
            var c = (x.children[0] as ParserRuleContext).children[0] as ParserRuleContext;
            var c2 = c.children[0] as ParserRuleContext;
            var p = c2.children[1] as ParserRuleContext;
	    if (p == undefined)
		return false;
	    var r = p.constructor.name === "ParametersAndQualifiersContext";
	    return r;
        } catch (e) {
        }
        return false;
    }
}

File: cpp/Cpp/CPP14ParserBase.cpp
#include "CPP14Parser.h"

using namespace antlr4;

bool CPP14ParserBase::IsPureSpecifierAllowed()
{
    try
    {
        auto x = this->getRuleContext(); // memberDeclarator
        auto c = x->children[0]->children[0];
        auto c2 = c->children[0];
        auto p = c2->children[1];
	if (p == nullptr) return false;
	return typeid(*p) == typeid(CPP14Parser::ParametersAndQualifiersContext);
    }
    catch (...)
    {
    }
    return false;
}
